Lib/test/test_urlparse.py

```diff
@@ -678,6 +734,38 @@ def test_urljoins_relative_base(self):
             self.checkJoin('http://a', scheme + '/w', 'http://a/w')
             self.checkJoin('http://a', scheme + '///w', 'http://a/w')
             self.checkJoin('http://a', scheme + 'w', 'http://a/w')
+            self.checkJoin('http://a', scheme + '../w', 'http://a/w')
+            self.checkJoin('http://a', scheme + './w', 'http://a/w')
+            self.checkJoin('http://a', scheme + '..//w', 'http://a/w')
+            self.checkJoin('http://a', scheme + './/w', 'http://a/w')
+            self.checkJoin('http://a', scheme + '..///w', 'http://a//w')
+            self.checkJoin('http://a', scheme + './//w', 'http://a//w')
+            self.checkJoin('http://a', scheme + '..', 'http://a')
+            self.checkJoin('http://a', scheme + '.', 'http://a')
+            self.checkJoin('http://a', scheme + '../', 'http://a')
+            self.checkJoin('http://a', scheme + './', 'http://a')
+            self.checkJoin('http://a', scheme + '..//', 'http://a/')
+            self.checkJoin('http://a', scheme + './/', 'http://a/')
+            self.checkJoin('http://a', scheme + '..///', 'http://a//')
+            self.checkJoin('http://a', scheme + './//', 'http://a//')
+
+        self.checkJoin('b/c', '', 'b/c')
+        self.checkJoin('b/c', '//', 'b/c')
```

Comment:
RFC 3986 carefully distinguishes between undefined and empty, and `//` has an empty authority, not undefined, so we should hit the `if defined(R.authority)` branch in §5.2.2. The result should be `//`.

Reply:
Yes, I know. I left them non-distinguished for compatibility. We will likely change this in a separate issue.